File: Azuga Assignment/URL Shortner Service/test/urlService.test.js
const { shortenUrl, getLongUrl } = require('../urlService/service');
global.urlMappings = {};

test('Valid URL Constistency', () => {
  const longUrl = 'https://google.com';
  const short1 = shortenUrl(longUrl);
  const short2 = shortenUrl(longUrl);
  expect(short1).toBe(short2);
});

test('Original URL from short code', () => {
  const longUrl = 'https://google.com';
  const shortCode = shortenUrl(longUrl);
  const original = getLongUrl(shortCode);
  expect(original).toBe(longUrl);
});

test('Check for unknown short code', () => {
  const result = getLongUrl('qwerty.123');
  expect(result).toBeUndefined();
});

test('Check for null input', () => {
    expect(() => shortenUrl(null)).toThrow();
});

test('Check for undefined input', () => {
expect(() => shortenUrl(undefined)).toThrow();
});

test('Check for very long URL input', () => {
const longUrl = 'https://google.com/' + 'a'.repeat(1000);
const shortCode = shortenUrl(longUrl);
expect(typeof shortCode).toBe('string');
expect(shortCode.length).toBeGreaterThan(0);
const original = getLongUrl(shortCode);
expect(original).toBe(longUrl);
});


test('Check for URLs with special characters', () => {
const longUrl = 'https://example.com/query?name=John&Doe&age=30';
const shortCode = shortenUrl(longUrl);
const original = getLongUrl(shortCode);
expect(original).toBe(longUrl);
});

test('Check for cases senstivity', () => {
const url1 = 'https://google.com/Page';
const url2 = 'https://google.com/page';
const short1 = shortenUrl(url1);
const short2 = shortenUrl(url2);
expect(short1).not.toBe(short2);
});

test('Check for service is persisted and retrievable after reload', () => {
const longUrl = 'https://test.com';
const shortCode = shortenUrl(longUrl);

global.urlMappings = {};// Simulate restart by clearing memory and reloading
const { getLongUrl } = require('../urlService/service');
const original = getLongUrl(shortCode);
expect(original).toBe(longUrl);
});

test('Check for similar URLs to be identify differently', () => {
const url1 = 'https://example.com/abc';
const url2 = 'https://example.com/def';
const short1 = shortenUrl(url1);
const short2 = shortenUrl(url2);
expect(short1).not.toBe(short2);
});
  

File: Azuga Assignment/url-shortener-java/src/test/java/UrlServiceTest.java
package service;


import service.UrlService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class UrlServiceTest {
    private UrlService urlService;

    @BeforeEach
    public void setUp() {
        urlService = new UrlService();
    }

    @Test
    public void testValidUrlConsistency() {
        String longUrl = "https://google.com";
        String short1 = urlService.shortenUrl(longUrl);
        String short2 = urlService.shortenUrl(longUrl);
        assertEquals(short1, short2);
    }

    @Test
    public void testOriginalUrlFromShortCode() {
        String longUrl = "https://google.com";
        String shortCode = urlService.shortenUrl(longUrl);
        String original = urlService.getLongUrl(shortCode);
        assertEquals(longUrl, original);
    }

    @Test
    public void testUnknownShortCode() {
        String result = urlService.getLongUrl("qwerty.123");
        assertNull(result);
    }

    @Test
    public void testNullInput() {
        assertThrows(IllegalArgumentException.class, () -> urlService.shortenUrl(null));
    }

    @Test
    public void testUndefinedInput() {
        assertThrows(IllegalArgumentException.class, () -> urlService.shortenUrl(""));
    }

    @Test
    public void testVeryLongUrlInput() {
        String longUrl = "https://google.com/" + "a".repeat(1000);
        String shortCode = urlService.shortenUrl(longUrl);
        assertNotNull(shortCode);
        assertTrue(shortCode.length() > 0);
        String original = urlService.getLongUrl(shortCode);
        assertEquals(longUrl, original);
    }

    @Test
    public void testUrlsWithSpecialCharacters() {
        String longUrl = "https://example.com/query?name=John&Doe&age=30";
        String shortCode = urlService.shortenUrl(longUrl);
        String original = urlService.getLongUrl(shortCode);
        assertEquals(longUrl, original);
    }

    @Test
    public void testCaseSensitivity() {
        String url1 = "https://google.com/Page";
        String url2 = "https://google.com/page";
        String short1 = urlService.shortenUrl(url1);
        String short2 = urlService.shortenUrl(url2);
        assertNotEquals(short1, short2);
    }

    @Test
    public void testSimilarUrlsIdentifiedDifferently() {
        String url1 = "https://example.com/abc";
        String url2 = "https://example.com/def";
        String short1 = urlService.shortenUrl(url1);
        String short2 = urlService.shortenUrl(url2);
        assertNotEquals(short1, short2);
    }
}